File: app-examples/example-jv-03adapters/src/main/java/foo/bar/example/foreadapters/feature/playlist/ImmutablePlaylistModel.java
package foo.bar.example.foreadapters.feature.playlist;

import java.util.ArrayList;
import java.util.List;

import co.early.fore.adapters.immutable.ImmutableListMgr;
import co.early.fore.core.Affirm;
import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.time.SystemTimeWrapper;

import static foo.bar.example.foreadapters.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource;

/**
 * Example model based on **immutable** list data
 *
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
public class ImmutablePlaylistModel extends ImmutableListMgr<Track> {

    private static final String LOG_TAG = ImmutablePlaylistModel.class.getSimpleName();

    private final Logger logger;

    public ImmutablePlaylistModel(SystemTimeWrapper systemTimeWrapper, WorkMode workMode, Logger logger) {
        super(systemTimeWrapper, workMode, logger);
        this.logger = Affirm.notNull(logger);
    }

    public void addNewTrack() {
        logger.i(LOG_TAG, "addNewTrack()");
        changeList(listCopy -> listCopy.add(new Track(generateRandomColourResource())));
    }

    public void removeTrack(int index) {
        logger.i(LOG_TAG, "removeTrack() " + index);
        checkIndex(index);
        changeList(listCopy -> listCopy.remove(index));
    }

    public void removeAllTracks() {
        logger.i(LOG_TAG, "removeAllTracks()");
        changeList(listCopy -> listCopy.clear());
    }

    public void increasePlaysForTrack(int index) {
        logger.i(LOG_TAG, "increasePlaysForTrack() " + index);
        checkIndex(index);
        changeList(listCopy -> listCopy.get(index).increasePlaysRequested());
    }

    public void decreasePlaysForTrack(int index) {
        logger.i(LOG_TAG, "decreasePlaysForTrack() " + index);
        checkIndex(index);
        changeList(listCopy -> listCopy.get(index).decreasePlaysRequested());
    }

    public void add5NewTracks() {
        logger.i(LOG_TAG, "add5NewTracks()");
        List<Track> newTracks = new ArrayList<>();
        for (int ii = 0; ii < 5; ii++) {
            newTracks.add(new Track(generateRandomColourResource()));
        }
        changeList(listCopy -> listCopy.addAll(0, newTracks));
    }

    public void remove5Tracks() {
        logger.i(LOG_TAG, "remove5Tracks()");
        if (getItemCount() > 4) {
            changeList(listCopy -> listCopy.subList(0, 5).clear());
        }
    }

    public void replaceTracks(List<Track> newTrackList) {
        logger.i(LOG_TAG, "replaceTracks()");
        if (getItemCount() > 4) {
            replaceList(() -> newTrackList);
        }
    }

    private void checkIndex(int index) {
        if (getItemCount() == 0) {
            throw new IndexOutOfBoundsException("tracklist has no items in it, can not get index:" + index);
        } else if (index < 0 || index > getItemCount() - 1) {
            throw new IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (getItemCount() - 1) + " not:" + index);
        }
    }
}

File: app-examples/example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/feature/playlist/immutable/ImmutablePlaylistModel.kt
package foo.bar.example.foreadapterskt.feature.playlist.immutable

import co.early.fore.adapters.Adaptable
import co.early.fore.adapters.immutable.Diffable
import co.early.fore.core.observer.Observable
import co.early.fore.kt.adapters.immutable.ImmutableListMgr
import co.early.fore.kt.core.logging.Logger
import foo.bar.example.foreadapterskt.feature.playlist.RandomStuffGeneratorUtil.generateRandomColourResource
import foo.bar.example.foreadapterskt.feature.playlist.RandomStuffGeneratorUtil.randomLong
import foo.bar.example.foreadapterskt.feature.playlist.Track
import java.util.ArrayList

/**
 * Example model based on **immutable** list data
 *
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
class ImmutablePlaylistModel (
        private val logger: Logger,
        private val listMgr: ImmutableListMgr<Track> = ImmutableListMgr(logger = logger)
) :
    Diffable by listMgr,
    Observable by listMgr,
    Adaptable<Track> by listMgr {

    fun removeTrack(index: Int) {
        logger.i("removeTrack() $index")
        checkIndex(index)
        listMgr.changeList {
            it.removeAt(index)
        }
    }

    fun removeAllTracks() {
        logger.i("removeAllTracks()")
        listMgr.changeList {
            it.clear()
        }
    }

    fun increasePlaysForTrack(index: Int) {
        logger.i("increasePlaysForTrack() $index")
        checkIndex(index)
        listMgr.changeList {
            it[index].increasePlaysRequested()
        }
    }

    fun decreasePlaysForTrack(index: Int) {
        logger.i("decreasePlaysForTrack() $index")
        checkIndex(index)
        listMgr.changeList {
            it[index].decreasePlaysRequested()
        }
    }

    fun addNTracks(n: Int) {
        logger.i("addNTracks() n:$n")

        val newTracks = ArrayList<Track>()
        for (ii in 0 until n) {
            newTracks.add(Track(generateRandomColourResource(), randomLong()))
        }

        listMgr.changeList {
            it.addAll(newTracks)
        }
    }

    fun removeNTracks(n: Int) {
        logger.i("removeNTracks() n:$n")
        if (getItemCount() > n - 1) {
            listMgr.changeList {
                it.subList(0, n).clear()
            }
        }
    }

    fun isEmpty(): Boolean {
        return !hasAtLeastNItems(1)
    }

    fun hasAtLeastNItems(n: Int): Boolean {
        return getItemCount() >= n
    }

    fun replaceTracks(newTrackList: List<Track>) {
        logger.i("replaceTracks()")
        listMgr.replaceList {
            newTrackList
        }
    }

    private fun checkIndex(index: Int) {
        if (isEmpty()) {
            throw IndexOutOfBoundsException("tracklist has no items in it, can not get index:$index")
        } else if (index < 0 || index > getItemCount() - 1) {
            throw IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (getItemCount() - 1) + " not:" + index)
        }
    }
}
